### ISSUE TITLE
kind create cluster creates same kubeconfig user name for each cluster

### ISSUE THREAD

**Issue by font** [kubernetes-sigs/kind#112]
When creating multiple clusters with `kind create cluster --name <name>` the kubeconfig for each cluster specifies the same user name of `kubernetes-admin`. This becomes a problem when trying to use multiple kubeconfigs in `KUBECONFIG` because the user name overlaps with each of the configs. For example:

```bash
$ export KUBECONFIG="$(kind get kubeconfig-path):$(kind get kubeconfig-path --name 2)"
$ kubectl --context=kubernetes-admin@kind-2 get all --all-namespaces
error: the server doesn't have a resource type "all"
```

Instead each auth user name should include the `--name` of the kind cluster when specified in the `kind create cluster --name <name>` command.

**Comment by BenTheElder** [comment]
this seems to be https://github.com/kubernetes/kubeadm/issues/416, looking into options...

**Comment by BenTheElder** [comment]
I [asked in #sig-cluster-lifecycle](https://kubernetes.slack.com/archives/C13J86Z63/p1542777463422000),  so far I don't think there's an answer for this. Following up...

**Comment by BenTheElder** [comment]
@neolit123 @fabriziopandini WDYT?
Afaict we'd have to provision our own user and matching kubeconfig.

**Comment by marckhouzam** [comment]
I've hit this problem previously while using multiple k8s clusters.  Turns out one can use the same username in all your clusters.  What needs to be tweaked in the kubeconfig file is changing the `name` field of the user to something unique and add the `username` field under `users.user` to be `kubernetes-admin`.

I'll try my hand at writing a PR for it.

**Comment by marckhouzam** [comment]
Ah, I see that it is kubeadm that generates the kubeconfig file and may not allow what kind needs for this.  I'll have a look at kubeadm some more.  But modifying the kubeconfig file after it is generated would be an option.

**Comment by marckhouzam** [comment]
Ok, I've got things working.  Here is what the new kubeconfig file looks like for a cluster named `mycluster` in case people want to manually make the changes for themselves:

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: REDACTED
    server: https://localhost:58039
  name: mycluster
contexts:
- context:
    cluster: mycluster
    user: kubernetes-admin-mycluster
  name: kubernetes-admin@mycluster
current-context: kubernetes-admin@mycluster
kind: Config
preferences: {}
users:
- name: kubernetes-admin-mycluster
  user:
    client-certificate-data: REDACTED
    client-key-data: REDACTED
    username: kubernetes-admin
```

I just need to get official approval from my employer to do the PR.  Should be a couple of days.

**Comment by discreet** [comment]
It seems that when I add `username` with accounts that use a token I get a multiple auth error:

eg: 
```
users:
  - name: foo
    user:
      token: bar
      username: baz
```

But when I add `username` to `users` it seems to work _until I change my context_ then it is removed from my kubeconfig.

eg:
```
users:
  - name: foo
    user:
      token: bar
    username: baz
```

**Comment by aojea** [comment]
/remove-lifecycle stale
This pops up from time to time in the channel

**Comment by BenTheElder** [comment]
see: https://github.com/kubernetes-sigs/kind/issues/850, proposing to solve this along with other changes

**Comment by BenTheElder** [comment]
fixing in #850. kind clusters will have unique entries. have this part implemented pretty cleanly.

**Comment by aojea** [comment]
/close
fixed by #1029 
thanks for the great feedback
